### DISCUSSION TITLE
Application Path

### DISCUSSION THREAD

**Discussion by dave-ccs** [ottowayi/pycomm3#238] [Q&A]
I'm trying to control an Invertek drive. According to the [product documentation](https://www.wolfautomation.com/media/pdf/ac-drives/invertek/odp/invertek-odp2-ethernet-ip-guide.pdf), pages 13-14, I need to connect to an application path of:
20 04 24 01 2C 01 2C 06

The 20 04 and 24 01 are standard class code and instance values. The 2C 01 and 2C 06 are O->T and T->O application paths. According to The CIP Networks Library Volume 1, section 3-5.5.1.10, this requires indicating O->T and T->O connection types. For this application, those are POINT2POINT.

As far as I can tell, pycomm3 doesn't (yet) support application connections or paths. Any suggestions on where to start and/or how to approach adding this functionality?

**Comment by ottowayi** [comment]
This isn't directly supported, but I think you can still make it work.  You would probably have to make your own version `generic_message` to send an unconnected message with your own request path.
Would probably need to create custom subclasses of the `GenericUnconnectedRequest/ResponsePacket` to encode the request path.  
  I think the `2C` is for a _Constructed Data Segment_, but the segment subtype is not one of the known ones (simple or ANSI extended symbol).  You can probably skip encoding the individual segments yourself and I think doing this should work: `request_path = PADDED_EPATH.encode([b'\x20\x04\x24\x01\x2C\x01\x2C\x06'])`.   

Unfortunately, I don't know if this will work and I'm unable to try it out.  I've been working on a full rewrite (albeit very slowly), so I apologize if you need to dig into those Packet classes.  That code is a bit of a mess and I'm working on making it much easier to implement custom functionality.  Right now, those classes handle (poorly) both the EtherNet/IP and CIP portions thru subclassing, which can be really confusing and hard to debug.  In the new version, something like this should be fairly straightforward to implement, but I have a long way to go and it will be a while before it's ready.

**Reply by dave-ccs** [reply]
Thanks for the quick response. I'll give it a try later this week. I will definitely post an update and example code if I get it to work.
